py/gc.c

```diff
@@ -805,27 +805,59 @@ void gc_dump_info(void) {
         (uint)info.num_1block, (uint)info.num_2block, (uint)info.max_block, (uint)info.max_free);
 }
 
-void gc_dump_alloc_table(void) {
+#if MICROPY_PY_MICROPYTHON_MEM_INFO_FRAGMENTATION
+// Power-of-two size classes. i.e. class[0] is how many single-block
+// allocations are possible. class[5] is how many 32-block allocations.
+#define NUM_FRAGMENTATION_CLASSES (20)
+
+// Given a number of contiguous blocks, figure out how many allocations of
+// each size class would fit in that many blocks.
+STATIC void gc_update_fragmentation_stats(size_t n, size_t *frag_classes) {
+    for (size_t c = 1, i = 0; c < n && i < NUM_FRAGMENTATION_CLASSES; c <<= 1, ++i) {
+        frag_classes[i] += n / c;
+    }
+}
+#endif
+
+#if MICROPY_PY_MICROPYTHON_MEM_INFO_BLOCKS || MICROPY_PY_MICROPYTHON_MEM_INFO_FRAGMENTATION
+void gc_dump_alloc_table(bool print_table, bool print_fragmentation) {
     GC_ENTER();
+    #if MICROPY_PY_MICROPYTHON_MEM_INFO_BLOCKS
     static const size_t DUMP_BYTES_PER_LINE = 64;
+    #endif
     #if !EXTENSIVE_HEAP_PROFILING
-    // When comparing heap output we don't want to print the starting
-    // pointer of the heap because it changes from run to run.
-    mp_printf(&mp_plat_print, "GC memory layout; from %p:", MP_STATE_MEM(gc_pool_start));
+    // Skip this when EXTENSIVE_HEAP_PROFILING is on because the starting
+    // pointer of the heap changes from run to run.
+    if (print_table) {
+        mp_printf(&mp_plat_print, "GC memory layout; from %p:", MP_STATE_MEM(gc_pool_start));
+    }
+    #endif
+    #if MICROPY_PY_MICROPYTHON_MEM_INFO_FRAGMENTATION
+    // How many consecutive free blocks.
+    size_t nfree = 0;
+    // Number of allocs that would succeed for 1, 2, 4, .. 2^n blocks.
+    size_t frag_classes[NUM_FRAGMENTATION_CLASSES] = {0};
```

Comment:
use `MICROPY_GC_STACK_ENTRY_TYPE` instead of `size_t` here

Reply:
Done.